Watchman.Discord/Areas/Initialization/Services/ServerScanningService.cs

```diff
@@ -20,58 +22,75 @@ public ServerScanningService(ICommandBus commandBus, MessagesHistoryService mess
-
+            var lastMessageId = 0UL;
             do
             {
-                messages = (await _messagesHistoryService.ReadMessagesAsync(server, channel, LIMIT, lastMessageId, goBefore: true)).ToList();
-                Serilog.Log.Information($"Channel: {channel.Name} read");
+                messages = ReadMessages(server, channel, LIMIT, lastMessageId);
                 if (messages.Count == 0)
                 {
                     break;
                 }
 
                 lastMessageId = messages.Last().Id;
                 await SaveMessages(messages, channel.Id);
-                Serilog.Log.Information($"Channel: {channel.Name} saved");
 
-            } while (messages.Count < LIMIT);
+                if (LastMessageIsOlderThanStartTime(messages, startTime.Value))
+                {
+                    break;
+                }
+
+            } while (messages.Count == LIMIT);
+
+            Log.Information($"Channel: {channel.Name} read and saved");
         }
 
-        private async Task SaveMessages(IEnumerable<Devscord.DiscordFramework.Services.Models.Message> messages, ulong channelId)
+        private async Task SaveMessages(IEnumerable<Message> messages, ulong channelId)
         {
             var convertedMessages = ConvertToMessages(messages);
             var command = new AddMessagesCommand(convertedMessages, channelId);
             await _commandBus.ExecuteAsync(command);
         }
 
-        private IEnumerable<Message> ConvertToMessages(IEnumerable<Devscord.DiscordFramework.Services.Models.Message> messages)
+        private IEnumerable<DomainModel.Messages.Message> ConvertToMessages(IEnumerable<Message> messages)
         {
             return messages.Select(x =>
             {
-                var builder = Message.Create(x.Request.OriginalMessage);
+                var builder = DomainModel.Messages.Message.Create(x.Request.OriginalMessage);
                 builder.WithAuthor(x.Contexts.User.Id, x.Contexts.User.Name);
                 builder.WithChannel(x.Contexts.Channel.Id, x.Contexts.Channel.Name);
                 builder.WithServer(x.Contexts.Server.Id, x.Contexts.Server.Name, x.Contexts.Server.Owner.Id, x.Contexts.Server.Owner.Name);
                 builder.WithSentAtDate(x.Request.SentAt);
                 return builder.Build();
             });
         }
+
+        private List<Message> ReadMessages(DiscordServerContext server, ChannelContext channel, int limit, ulong lastMessageId = 0)
+        {
+            var messages = lastMessageId == 0
+                ? _messagesHistoryService.ReadMessagesAsync(server, channel, limit).Result.ToList()
+                : _messagesHistoryService.ReadMessagesAsync(server, channel, limit, lastMessageId, goBefore: true).Result.ToList();
+
+            return messages;
+        }
+
+        private bool LastMessageIsOlderThanStartTime(IEnumerable<Message> messages, DateTime startTime)
```

Comment:
chyba lepiej wrzucić tam po prostu to messages.Last().Request.SentAt < startTime
bo to dość prosty kod który jasno mówi co i jak
a przez dodatkową metode z jednej linijki robi nam sie kilka

Reply:
Ale używana jest dwa razy, dlatego zrobiłem jako metodę. Myślisz, że lepiej wkleić w oba miejsca?

Reply:
hmm
dobra, zostawmy to